controllers/operatorpolicy_controller.go

```diff
@@ -133,9 +146,96 @@ func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// FUTURE: more resource checks
 
+	// handle catalogsource
+	if err := r.handleCatalogSource(ctx, policy); err != nil {
+		OpLog.Error(err, "Error handling CatalogSource")
+
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
 
+func (r *OperatorPolicyReconciler) handleCatalogSource(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+) error {
+	watcher := opPolIdentifier(policy.Namespace, policy.Name)
+
+	desiredSub, err := buildSubscription(policy)
+	if err != nil {
+		return fmt.Errorf("error building Subscription: %w", err)
+	}
+
+	// Retrieve Subscription
+	foundSub, err := r.DynamicWatcher.Get(watcher, subscriptionGVK, desiredSub.Namespace, desiredSub.Name)
+	if err != nil {
+		return fmt.Errorf("error getting Subscription: %w", err)
+	}
+
+	if foundSub == nil {
+		// Subscription not found, check the CatalogSource object
+		foundCatalogSrc, err := r.DynamicWatcher.Get(watcher, catalogSrcGVK,
+			desiredSub.Spec.CatalogSourceNamespace, desiredSub.Spec.CatalogSource)
+		if err != nil {
+			return fmt.Errorf("error getting CatalogSource: %w", err)
+		}
+
+		// Conditionally emit health event using the defined status template
+		isMissing := foundCatalogSrc == nil
+
+		err = r.updateStatus(ctx, policy, catalogSourceFindCond(isMissing), catalogSourceObj(desiredSub, isMissing))
+		if err != nil {
+			return fmt.Errorf("error updating the status for CatalogSource when Subscription DNE on cluster: %w", err)
+		}
+
+		return nil
+	}
+
+	condList, found, err := unstructured.NestedSlice(foundSub.Object, "status", "conditions")
+	if !found {
+		return fmt.Errorf("field does not exist in object: %w", err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error decoding object into array type: %w", err)
+	}
+
+	// Iterate over all the conditions to determine the state of the CatalogSource
+	for _, condition := range condList {
+		condMap, ok := condition.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("conditions field is of the wrong type")
```

Comment:
This makes me realize that we might need to keep track of 2 kinds of errors: this one here is something that re-reconciling can not fix. So if this occurs, any `handle` functions after it won't be run, which means the status might always be incomplete. That's probably not ideal, I think it's something we'll need to deal with eventually.

On the other hand, it's possible to have an error when reading from the API server, which does just need a retry, so I think re-reconciling in that case is fine.

Anyway, I think this is fine for this PR, I was just mentioning it for awareness.

Reply:
Yea, I think we probably want some type of error message that can be displayed/logged, but not re-reconcile when this type of error happens.